Update C++ rules (#3148)
SonarSource/rspec: rules/S5502/cfamily/rule.adoc

== Why is this an issue?

In a statement, the order of evaluation of sub-expressions (e.g., the arguments of a function call) is not totally specified. This means the compiler can even interleave the evaluation of these sub-expressions, especially for optimization purposes.

> If you have several resource allocations in one statement, and the first succeeds while the second fails and throws an exception, the first allocated resource can leak. The classical mitigation for this issue is to use an RAII (_Resource Acquisition Is Initialization_) manager to wrap the raw resource, but since the execution order is not specified, you may not have the time to do this.

Reviewer: I don't understand this part: "you may not have the time to do this". Maybe this is clearer?
```suggestion
If you have several resource allocations in one statement, and the first succeeds while the second fails and throws an exception, the first allocated resource can leak. The classical mitigation for this issue is to use an RAII (_Resource Acquisition Is Initialization_) manager to wrap the raw resource. Yet, this solution may not be sufficient since the execution order is not specified.
```

> It is possible to write code that contains several allocations and will still behave correctly (it has even been made easier in {cpp}17, where the evaluation order has been made more strict). However, it requires expert-level knowledge of the language. It is simpler and more future-proof to simply avoid using several allocations in a single statement.

Reviewer: Present tense is better IMO. 
```suggestion
It is possible to write code that contains several allocations and still behaves correctly. {cpp}17 made this even easier since the evaluation order rules are more strict. However, it requires expert-level knowledge of the language. It is simpler and more future-proof to simply avoid using several allocations in a single statement.
```

> It is possible to write code that contains several allocations and will still behave correctly (it has even been made easier in {cpp}17, where the evaluation order has been made more strict). However, it requires expert-level knowledge of the language. It is simpler and more future-proof to simply avoid using several allocations in a single statement.

Reviewer: Here we say "better not to have the problem in the first place" but in the example, we don't put this forward at all. I wonder if we should suggest an alternative for explicitness that removes the problem altogether. WDYT?

Author: I added a more straightforward solution: https://github.com/SonarSource/rspec/pull/3148/commits/6e24e4e4b2eea2d50fa66721cbfb0ede62a6c993
WDYT?

Reviewer: Looks good!

> === Noncompliant code example

Reviewer: Here too, I find the `using` statement more of a distraction.

Author: Fixed in https://github.com/SonarSource/rspec/pull/3148/commits/6e24e4e4b2eea2d50fa66721cbfb0ede62a6c993

> == Resources

Reviewer: Above, the comment "Compliant: no resource in the same expression statement" is puzzling. I think it should say "no two resources are allocated in the same expression statement" but this may be overcomplicated...

Author: Modified in https://github.com/SonarSource/rspec/pull/3148/commits/6e24e4e4b2eea2d50fa66721cbfb0ede62a6c993. Sounds clearer?

Reviewer: Perfect
